packages/components/src/higher-order/with-focus-outside/index.js

```diff
@@ -1,142 +1,35 @@
-			this.normalizeButtonFocus = this.normalizeButtonFocus.bind( this );
-		}
-
-		componentWillUnmount() {
-			this.cancelBlurCheck();
-		}
-
-		bindNode( node ) {
-			if ( node ) {
-				this.node = node;
-			} else {
-				delete this.node;
-				this.cancelBlurCheck();
-			}
-		}
-
-		queueBlurCheck( event ) {
-			// React does not allow using an event reference asynchronously
-			// due to recycling behavior, except when explicitly persisted.
-			event.persist();
-
-			// Skip blur check if clicking button. See `normalizeButtonFocus`.
-			if ( this.preventBlurCheck ) {
-				return;
-			}
-
-			this.blurCheckTimeout = setTimeout( () => {
-				// If document is not focused then focus should remain
-				// inside the wrapped component and therefore we cancel
-				// this blur event thereby leaving focus in place.
-				// https://developer.mozilla.org/en-US/docs/Web/API/Document/hasFocus.
-				if ( ! document.hasFocus() ) {
-					event.preventDefault();
-					return;
-				}
-				if ( 'function' === typeof this.node.handleFocusOutside ) {
-					this.node.handleFocusOutside( event );
-				}
-			}, 0 );
-		}
-
-		cancelBlurCheck() {
-			clearTimeout( this.blurCheckTimeout );
-		}
-
-		/**
-		 * Handles a mousedown or mouseup event to respectively assign and
-		 * unassign a flag for preventing blur check on button elements. Some
-		 * browsers, namely Firefox and Safari, do not emit a focus event on
-		 * button elements when clicked, while others do. The logic here
-		 * intends to normalize this as treating click on buttons as focus.
-		 *
-		 * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
-		 *
-		 * @param {MouseEvent} event Event for mousedown or mouseup.
-		 */
-		normalizeButtonFocus( event ) {
-			const { type, target } = event;
-
-			const isInteractionEnd = includes(
-				[ 'mouseup', 'touchend' ],
-				type
-			);
-
-			if ( isInteractionEnd ) {
-				this.preventBlurCheck = false;
-			} else if ( isFocusNormalizedButton( target ) ) {
-				this.preventBlurCheck = true;
-			}
-		}
-
-		render() {
-			// Disable reason: See `normalizeButtonFocus` for browser-specific
-			// focus event normalization.
-
-			/* eslint-disable jsx-a11y/no-static-element-interactions */
-			return (
-				<div
-					onFocus={ this.cancelBlurCheck }
-					onMouseDown={ this.normalizeButtonFocus }
-					onMouseUp={ this.normalizeButtonFocus }
-					onTouchStart={ this.normalizeButtonFocus }
-					onTouchEnd={ this.normalizeButtonFocus }
-					onBlur={ this.queueBlurCheck }
-				>
-					<WrappedComponent ref={ this.bindNode } { ...this.props } />
-				</div>
-			);
-			/* eslint-enable jsx-a11y/no-static-element-interactions */
-		}
-	};
-}, 'withFocusOutside' );
+import { useFocusOutside } from '../../utils/hooks';
+
+export default createHigherOrderComponent(
+	( WrappedComponent ) => ( props ) => {
+		const [ handleFocusOutside, setHandleFocusOutside ] = useState();
```

Comment:
Why  do we need this state at all, why  not just pass the callback to `useFocusOutside`?

Reply:
Ok, long story:

Passing the callback works if it's wrapped in a closure in `withFocusOutside` like in your PR:
```jsx
useFocusOutside( ( event ) => node.current?.handleFocusOutside( event ) );
```

However, I went for `useState` because of this method:
https://github.com/WordPress/gutenberg/blob/454c7085d23c155ebbdcb1621b8881fba5311552/packages/components/src/higher-order/with-focus-outside/index.js#L58-L65

That seems to say when `node` becomes a falsey value `cancelBlurCheck` should be triggered.

So the idea of using `state` is to emulate that, the state gets unset when the node isn't present and `cancelBlurCheck` is triggered in the other bit of code that you queried here - https://github.com/WordPress/gutenberg/pull/27369#discussion_r538119820

If I go with the option of using the closure mentioned at the top of this comment, it means there's always an `onFocusOutside` callback provided to the hook so it won't be reactive to the node being `falsey`.

TBH, I find the use of `cancelBlurCheck` in `bindNode` quite hard to reason about. Is there likely to be a blur check in flight when the node is removed?

Reply:
Did you try to remove it and see whether any test fail?

Reply:
Just looking now, none of the unit tests fail and I didn't see any issues in manual testing. I can push the change to see if the e2e tests pass?